### ISSUE TITLE
Auto-router uses inactive layers if via cost set to 45 or lower

### ISSUE THREAD

**Issue by mgroz** [freerouting/freerouting#230]
### Please provide as many details as you can
* What version of freerouting are you using?
freerouting plugin 1.8.0 and stand alone program 1.8 (2023-5-22)
* On what platform did you run it? Windows, Linux, MacOS? 32-bit or 64-bit?
Windows 10 (64 bit)
* What EDA do you use? KiCad, KiCad with freerouting plugin, EAGLE, EasyEDA, other? What is its exact version?
KiCad 7.0.7,
* What types of files are you working with? Can you attach the project files (.pro, .kicad_pcb, etc) or the exported DSN file?
KiCad 7: I sent you the files to both info@freerouting.app and notifications@github.com
* Did you use any command line arguments?
No
* Can you post a screenshot of the problem?
* See below, green routes on inner 1 power layer
![image](https://github.com/freerouting/freerouting/assets/119752187/effbc860-db9e-45e3-829b-bec4b3481b7d)
The below is my board stackup.
![image](https://github.com/freerouting/freerouting/assets/119752187/c9b71234-c585-4213-bd01-e304d316d432)

As shown above, In1.Cu & In2.Cu (inner layers) are assigned as power planes. The below picture shows these layers are not even listed by freerouting so why are they being routed on.
![image](https://github.com/freerouting/freerouting/assets/119752187/9ec49c87-f2a5-4213-9544-8042bc17e1d6)

* Can you include the latest, related log entries from the .log file?
See above, I've sent them twice.
* Please list the detailed steps to reproduce the problem
This problem is with both the plugin and 1.8 stand alone specctra software. I just kick off the software then make the parameter changes (see picture), and when it comes back into KiCad the routes are present on power layer.

* What would you expect freerouting to do in this situation?
Not route on layers that are not checked to be routed on.

* Is fixing this issue critical for your work?
yes

Thanks

**Comment by andrasfuchs** [comment]
Hey @mgroz, thanks for the report!

I didn't get the files, would you mind attaching them here as a zip file?

**Comment by andrasfuchs** [comment]
I got the files, thank you, and thank you for the kind words.

As far as I can tell the exported DSN file has all the 4 layers, but freerouting somewhat ignores the ones that are defined as `power`. Strangely though they are only ignored on the user interface, not in the routing algorithm (as you reported it).

A better solution would be to display the power layers too in both Parameter / Select and Parameter / Autoroute windows, but disable auto-routing for them by default.

**Comment by andrasfuchs** [comment]
I added the non-signal layer to the two dialogs:
![image](https://github.com/freerouting/freerouting/assets/910321/5d1edbe9-1e39-4981-b196-bd40ce30e549)
![image](https://github.com/freerouting/freerouting/assets/910321/bf33cad7-e3c4-4de2-b56f-d80845cdfbc5)

After the routing your board looks ok now:
![image](https://github.com/freerouting/freerouting/assets/910321/78c8201b-63f6-44bd-8631-5a7c9dff9f19)

**Comment by andrasfuchs** [comment]
It isn't available as KiCad plugin yet, but if everything goes as planned, you will see the new version next week.

Until then you can test it by using the snapshot build: https://github.com/freerouting/freerouting/releases/tag/SNAPSHOT

**Comment by andrasfuchs** [comment]
I tried to reproduce the problem again, and I found something interesting: when I export the DSN file from KiCad and I open it in a text editor you can find a few, already wired nets:

![image](https://github.com/freerouting/freerouting/assets/910321/02d4ded5-84f9-4b59-a18e-5df4b7da6d1d)

As you can see there are already entries with In1.Cu which means that they were set in KiCad, not Freerouting, so it's going to have some wires in the end after the autorouting, no matter what Freerouting does.

If you - on the other hand - run Freerouting, open the DSN file, and choose "Delete all tracks and vias", and then run the autorouter you will not have any wires on the power layers.

![image](https://github.com/freerouting/freerouting/assets/910321/d5867f8b-9d85-44c3-88f1-22c16efb509b)

If I change the layer types from `power` to `signal`, the auto-router settings were working as expected.

I made a small change to the UI: autorouting is forced to be disabled for non-signal layers.

Could you confirm that if you use the "Delete all tracks and vias" function with [the latest snapshot build](https://github.com/freerouting/freerouting/releases/tag/SNAPSHOT) then the auto-routing works fine?

**Comment by andrasfuchs** [comment]
Thanks for the details again, I'll look into it today.

The `NoClassDefFoundError` is a strange one, because it should only cause any problems if you had an older Java runtime, but you clearly have a Java 20. It's not a custom or stripped down version of the Java runtime, is it?

It will not cause any issues (other than the popup window), it just tries to check for a new version at startup by using the HttpClient class.

**Comment by andrasfuchs** [comment]
It seems like I can't force freerouting to wire on not-allowed layers, even if I change their type from power to mixed/signal.

If I export your board from KiCad, I get a DSN file that has four layers: signal, power, power, signal. If I start the auto-router after this, I get wires only on the two signal layers (as expected).

If I change their type to: signal, signal, signal, signal (mixed layers in KiCad are saved as "signal" in the DSN), then all layers are used (as expected).

I tried multiple combinations of enabling layers, all seemed fine, so I must miss something important I think.

I know you gave me many details already, and sorry to ask again, but could you give me step by step instructions to reproduce this problem consistently? This is one of the last issues I plan to fix before the next release.

Your JRE installation path seems odd indeed, this might be a problem, I'm not sure. The curious thing is that the HttpClient error message you have supposed to be happening with JRE version 9 and below. 
It's very unlikely, but is it possible that you have multiple version of Java installed, and because of the path issue Freerouting is run by an older version?

**Comment by andrasfuchs** [comment]
Yes! I think we finally got it: if we change to `via cost` in _Settings / Auto-router / Detailed Settings_ dialog to 25 (and hit enter to apply) and then we run the autorouter, we will get some wires in the inner layers, even if they are not active in the _Settings / Auto-router_.

We don't need to have the _Fanout_ or _Postroute_ options checked.

According to my tests the bug is present if I change the via cost to 45 or lower,  but it doesn't occur if I set it to 46 or higher. The interesting side effect of this change is that the number of auto-router passes increases significantly with the lower settings, 5 (cost=46) vs 210 (cost=45) in my case.

The exact location of the bug isn't obvious for me at the moment, it will need more investigation.

Thank you @mgroz for the continuous reports and tests!
